File: php/leet_code/3075.php
<?php
/**
 * Copyright ©2012-2024 JerodG <https://github.com/jerodg/>
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the
 * Server Side Public License (SSPL) as published by MongoDB, Inc., either version 1 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the SSPL
 * for more details.
 *
 * The above copyright notice and this permission notice shall be included in all copies or
 * substantial portions of the Software. You should have received a copy of the SSPL along with this
 * program. If not, see <https://www.mongodb.com/licensing/server-side-public-license>.
 */

declare(strict_types=1);

/**
 * Class Solution
 *
 * This class provides a solution for calculating the maximum happiness sum.
 */
class Solution
{
    /**
     * Calculate the maximum happiness sum.
     *
     * This method calculates the maximum happiness sum by sorting the happiness array in ascending order,
     * then decrementing the highest values in the array by 1 for each iteration up to k times.
     * The decremented values are then added to the output if they are greater than 0.
     * The process stops when the decremented value is less than or equal to 0 or when the number of
     * iterations reaches k.
     *
     * @param array $happiness The array of happiness values.
     * @param int $k The number of times to decrement the highest values in the array.
     * @return int The maximum happiness sum.
     */
    public function maximumHappinessSum(array $happiness, int $k): int
    {
        sort($happiness);
        $dec = 0;
        $out = 0;
        for ($i=0; $i<$k; $i++) {
            $h = array_pop($happiness) - $dec;
            $dec++;
            if ($h>0) {
                $out += $h;
            } else {
                break;
            }
        }
        return $out;
    }
}

File: python/leet_code/3075.py
"""Copyright ©2010-2025 JerodG <https://github.com/jerodg/>

This program is free software: you can redistribute it and/or modify it under the terms of the
Server Side Public License (SSPL) as published by MongoDB, Inc., either version 1 of the License,
or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the SSPL
for more details.

The above copyright notice and this permission notice shall be included in all copies or
substantial portions of the Software. You should have received a copy of the SSPL along with this
program. If not, see <https://www.mongodb.com/licensing/server-side-public-license>.
"""


class Solution:
    """Provides solutions for the maximum happiness sum problem.

    Implements algorithmic solutions for calculating the maximum
    happiness achievable when giving gifts to children with various constraints.
    """

    @staticmethod
    def maximumHappinessSum(happiness: list[int], k: int) -> int:
        """Calculates the maximum possible sum of happiness after giving gifts to children.

        The happiness of each child is reduced by their position in the gift-giving order
        (0-indexed). A gift is only given if the resulting happiness is positive.

        Args:
            happiness: A list of integers representing each child's happiness value.
            k: The maximum number of children to give gifts to.

        Returns:
            The maximum possible sum of happiness after giving gifts to at most k children.

        Examples:
            >>> Solution().maximumHappinessSum([1, 2, 3], 2)
            4
            >>> Solution().maximumHappinessSum([10, 5, 8], 3)
            20
        """
        # Sort children by happiness in descending order to prioritize the happiest children
        children = sorted(happiness, reverse=True)

        # Optimization: If the k-th happiest child still has positive happiness
        # after decrementing (k-1) times, we can use a formula to calculate the total
        if children[k - 1] >= k - 1:
            # Sum of first k children minus the sum of decrements (arithmetic sequence)
            return sum(children[:k]) - ((0 + k - 1) * k // 2)

        res = 0

        # Process each child in order of decreasing happiness
        for i, h in enumerate(children[:k]):
            # Skip children who would have zero or negative happiness after decrement
            if h - i <= 0:
                break
            # Add adjusted happiness value to the result
            res += h - i

        return res
